wgsl/extract-grammar.py

```diff
@@ -419,6 +425,78 @@ def not_token_only(value):
     grammar_package.write('    "main": "bindings/node"\n')
     grammar_package.write('}\n')
 
+# External scanner for nested block comments
+# See: https://github.com/tree-sitter/tree-sitter-rust/blob/master/src/scanner.c
+
+os.makedirs(os.path.join(grammar_path, "src"), exist_ok=True)
+with open(os.path.join(grammar_path, "src", "scanner.c"), "w") as external_scanner:
+    external_scanner.write(r"""
+#include <tree_sitter/parser.h>
+#include <wctype.h>
+
+enum TokenType {
+  BLOCK_COMMENT,
+};
+
+void *tree_sitter_wgsl_external_scanner_create() { return NULL; }
+void tree_sitter_wgsl_external_scanner_destroy(void *p) {}
+void tree_sitter_wgsl_external_scanner_reset(void *p) {}
+unsigned tree_sitter_wgsl_external_scanner_serialize(void *p, char *buffer) { return 0; }
+void tree_sitter_wgsl_external_scanner_deserialize(void *p, const char *b, unsigned n) {}
+
+static void advance(TSLexer *lexer) {
+  lexer->advance(lexer, false);
+}
+
+bool tree_sitter_wgsl_external_scanner_scan(void *payload, TSLexer *lexer,
+                                            const bool *valid_symbols) {
+  while (iswspace(lexer->lookahead)) lexer->advance(lexer, true);
+
+  if (lexer->lookahead == '/') {
+    advance(lexer);
+    if (lexer->lookahead != '*') return false;
+    advance(lexer);
+
+    bool after_star = false;
+    unsigned nesting_depth = 1;
+    for (;;) {
+      switch (lexer->lookahead) {
+        case '\0':
```

Comment:
So this is interesting.  I had not thought about the NUL character in source code.
UTF-8 allows it.  We would need a decision in the group about whether WGSL source can contain it.

I'll file an issue to track that.
I'll approve this and we can revisit if the group decides to allow NUL characters in comments.

Reply:
@dneto0 thanks for pointing it out, a spec discussion will definitely benefit.

Reply:
I just read tree-sitter's lexer.   I  think this actually signals the end of input.

Internally, when the lexer hits end of input, it sets TSLexer.lookahead to '\0' and the internal "lookahead_size" to 1.
See https://github.com/tree-sitter/tree-sitter/blob/67de9435b109f9bd8bf5957d11ff13161966d262/lib/src/lexer.c#L150  

This doesn't seem to be documented.  :-(